### ISSUE TITLE
Exceptions: index out of bound and CIGAR issue related to realignment

### ISSUE THREAD

**Issue by cbrueffer** [AstraZeneca-NGS/VarDictJava#243]
Hi,

I'm experiencing the following exceptions with VarDict-Java 1.6.0 in HISAT 2.1.0 aligned data and the following invocation: `vardict-java -G /genomes/Hsapiens/hg38-mut-cb/seq/hg38-mut-cb.fa -f 0.02 -c 1 -S 2 -E 3 -g 4 --nosv -b test-malformed-cigar-and-string-eor.bam -R chr4:25001600-25001607`. The issue goes away when disabling realignment (`-k 0`).

```
There was Exception while processing cigar on 25001538 16M1I1M1I2M1D3M1I50M on region chr4:25001600-25001607. The processing will be continued from the next cigar.
java.lang.StringIndexOutOfBoundsException: String index out of range: 90
        at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:47)
        at java.base/java.lang.String.charAt(String.java:693)
        at com.astrazeneca.vardict.variations.VariationUtils.isHasAndNotEquals(VariationUtils.java:498)
        at com.astrazeneca.vardict.modules.CigarModifier.captureMisSoftly3Mismatches(CigarModifier.java:320)
        at com.astrazeneca.vardict.modules.CigarModifier.modifyCigar(CigarModifier.java:178)
        at com.astrazeneca.vardict.modules.CigarParser.parseCigar(CigarParser.java:270)
        at com.astrazeneca.vardict.modules.CigarParser.process(CigarParser.java:89)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
        at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094)
        at com.astrazeneca.vardict.modes.AbstractMode.pipeline(AbstractMode.java:48)
        at com.astrazeneca.vardict.modes.SimpleMode.processBamInPipeline(SimpleMode.java:103)
        at com.astrazeneca.vardict.modes.SimpleMode.notParallel(SimpleMode.java:45)
        at com.astrazeneca.vardict.VarDictLauncher.start(VarDictLauncher.java:65)
        at com.astrazeneca.vardict.Main.main(Main.java:15)
There was Exception while processing record on NS500281:178:HNYVKBGXX:3:22611:10789:14832 on region chr4:25001600-25001607. The processing will be continued from the next record.
java.lang.IllegalArgumentException: Malformed CIGAR string: 31M1I-8M1I50M
        at htsjdk.samtools.TextCigarCodec.decode(TextCigarCodec.java:69)
        at com.astrazeneca.vardict.modules.CigarParser.parseCigar(CigarParser.java:272)
        at com.astrazeneca.vardict.modules.CigarParser.process(CigarParser.java:89)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
        at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094)
        at com.astrazeneca.vardict.modes.AbstractMode.pipeline(AbstractMode.java:48)
        at com.astrazeneca.vardict.modes.SimpleMode.processBamInPipeline(SimpleMode.java:103)
        at com.astrazeneca.vardict.modes.SimpleMode.notParallel(SimpleMode.java:45)
        at com.astrazeneca.vardict.VarDictLauncher.start(VarDictLauncher.java:65)
        at com.astrazeneca.vardict.Main.main(Main.java:15)
```

The exceptions are non-fatal and variant calling continues.

A minimal BAM file showing the problems: https://lu.box.com/s/w1cq3q2t4k6h2vmn7s5fgqlz52c4znrw

**Comment by PolinaBevad** [comment]
Christian, hello!

Thank you for the detailed description and BAM file for this issue! 

I looked at the data and found that there is a sequence in this record with repeat bases that can be recognized as a long shift to the left end, and we didn't consider such a case before, so we have negative values for matched part after modifying cigar. 

I've prepared the fix for this in PR #244. 
Thank you!
